File: common/src/main/java/dev/chasem/cobblemonextras/commands/PokeOdds.java
package dev.chasem.cobblemonextras.commands;

import com.cobblemon.mod.common.Cobblemon;
import com.cobblemon.mod.common.api.storage.party.PlayerPartyStore;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.DoubleArgumentType;
import com.mojang.brigadier.arguments.FloatArgumentType;
import com.mojang.brigadier.arguments.IntegerArgumentType;
import com.mojang.brigadier.context.CommandContext;
import dev.chasem.cobblemonextras.CobblemonExtras;
import dev.chasem.cobblemonextras.permissions.CobblemonExtrasPermissions;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;

import static dev.chasem.cobblemonextras.util.PokemonUtility.getHoverText;
import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class PokeOdds {

    public void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        // Register /pokeodds to get the current shiny rate no permission required, and /pokeodds setRate <rate> to set the shiny rate, requires permission
        dispatcher.register(literal("pokeodds")
                .then(literal("setRate")
                        .requires(source -> CobblemonExtrasPermissions.checkPermission(source, CobblemonExtras.permissions.POKEODDS_PERMISSION))
                        .then(argument("rate", FloatArgumentType.floatArg(0.0F, 100.0F))
                                .executes(ctx -> setRate(ctx, FloatArgumentType.getFloat(ctx, "rate")))))
                .executes(this::execute));
    }

    private int setRate(CommandContext<ServerCommandSource> ctx, float rate) {
        ctx.getSource().sendMessage(Text.literal("The shiny rate has been set to: ").formatted(Formatting.GOLD)
                .append(Text.literal(String.valueOf(rate)).formatted(Formatting.AQUA)));
        Cobblemon.config.setShinyRate(rate);
        return 1;
    }

    private int execute(CommandContext<ServerCommandSource> ctx) {
        ctx.getSource().sendMessage(Text.literal("The current shiny rate is: ").formatted(Formatting.GOLD)
                .append(Text.literal(String.valueOf(Cobblemon.config.getShinyRate())).formatted(Formatting.AQUA)));
        return 1;
    }



}

File: common/src/main/kotlin/dev/chasem/cobblemonextras/commands/PokeOdds.kt
package dev.chasem.cobblemonextras.commands

import com.cobblemon.mod.common.Cobblemon
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.arguments.FloatArgumentType
import com.mojang.brigadier.context.CommandContext
import dev.chasem.cobblemonextras.CobblemonExtras
import dev.chasem.cobblemonextras.config.CobblemonExtrasConfig
import dev.chasem.cobblemonextras.permissions.CobblemonExtrasPermissions
import net.minecraft.ChatFormatting
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.Commands
import net.minecraft.network.chat.Component
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import com.google.gson.JsonObject
import com.google.gson.JsonParser

class PokeOdds {
    fun register(dispatcher: CommandDispatcher<CommandSourceStack>) {
        dispatcher.register(
            Commands.literal("pokeodds")
                // anyone can read the current rate
                .executes { ctx -> execute(ctx) }
                .then(
                    Commands.literal("setRate")
                        // Allow if OP level >= 2, or the explicit permission is granted, or it's the console
                        .requires { src ->
                            val isConsole = src.entity == null
                            val isOp = src.hasPermission(2) // vanilla admin level gate
                            val hasNode = CobblemonExtrasPermissions.checkPermission(
                                src, CobblemonExtras.permissions.POKEODDS_PERMISSION
                            )
                            isConsole || isOp || hasNode
                        }
                        .then(
                            Commands.argument("rate", FloatArgumentType.floatArg(1.0f, 10000.0f))
                                .executes { ctx ->
                                    setRate(ctx, FloatArgumentType.getFloat(ctx, "rate"))
                                }
                        )
                )
        )
    }

    private fun setRate(ctx: CommandContext<CommandSourceStack>, rate: Float): Int {
        Cobblemon.config.shinyRate = rate
        
        // Save the Cobblemon config to persist changes across restarts
        try {
            saveCobblemonConfig(rate)
            ctx.source.sendSystemMessage(
                Component.literal("The shiny rate has been set to: ").withStyle(ChatFormatting.GOLD)
                    .append(Component.literal(rate.toString()).withStyle(ChatFormatting.AQUA))
                    .append(Component.literal(" (saved to config)").withStyle(ChatFormatting.GREEN))
            )
        } catch (e: Exception) {
            ctx.source.sendSystemMessage(
                Component.literal("The shiny rate has been set to: ").withStyle(ChatFormatting.GOLD)
                    .append(Component.literal(rate.toString()).withStyle(ChatFormatting.AQUA))
                    .append(Component.literal(" (WARNING: failed to save to config)").withStyle(ChatFormatting.RED))
            )
            CobblemonExtras.getLogger().error("Failed to save Cobblemon config after setting shiny rate", e)
        }
        
        return 1
    }

    private fun saveCobblemonConfig(newShinyRate: Float) {
        val configFileLoc = System.getProperty("user.dir") + File.separator + "config" + File.separator + "cobblemon" + File.separator + "main.json"
        val configFile = File(configFileLoc)
        
        if (!configFile.exists()) {
            CobblemonExtras.getLogger().warn("Cobblemon config file not found at: $configFileLoc")
            return
        }
        
        try {
            // Read the existing config
            val fileReader = FileReader(configFile)
            val configJson = JsonParser.parseReader(fileReader).asJsonObject
            fileReader.close()
            
            // Update the shiny rate directly at the root level
            configJson.addProperty("shinyRate", newShinyRate)
            
            // Write the updated config back to file
            val fileWriter = FileWriter(configFile)
            CobblemonExtrasConfig.GSON.toJson(configJson, fileWriter)
            fileWriter.flush()
            fileWriter.close()
            
            CobblemonExtras.getLogger().info("Successfully saved shiny rate $newShinyRate to Cobblemon config")
        } catch (e: Exception) {
            CobblemonExtras.getLogger().error("Failed to save Cobblemon config", e)
            throw e
        }
    }

    private fun execute(ctx: CommandContext<CommandSourceStack>): Int {
        ctx.source.sendSystemMessage(
            Component.literal("The current shiny rate is: ").withStyle(ChatFormatting.GOLD)
                .append(Component.literal(Cobblemon.config.shinyRate.toString()).withStyle(ChatFormatting.AQUA))
        )
        return 1
    }
}
